File: libs/moqtail-ts/src/model/control/subscribe_announces.ts
import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Tuple } from '../common/tuple'
import { KeyValuePair } from '../common/pair'
import { ControlMessageType } from './constant'
import { LengthExceedsMaxError } from '../error/error'

export class SubscribeAnnounces {
  constructor(
    public readonly requestId: bigint,
    public readonly trackNamespacePrefix: Tuple,
    public readonly parameters: KeyValuePair[],
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.SubscribeAnnounces
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.SubscribeAnnounces)
    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putTuple(this.trackNamespacePrefix)
    payload.putVI(this.parameters.length)
    for (const param of this.parameters) {
      payload.putKeyValuePair(param)
    }
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('SubscribeAnnounces::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): SubscribeAnnounces {
    const requestId = buf.getVI()
    const trackNamespacePrefix = buf.getTuple()
    const paramCount = buf.getNumberVI()
    const parameters: KeyValuePair[] = new Array(paramCount)
    for (let i = 0; i < paramCount; i++) {
      parameters[i] = buf.getKeyValuePair()
    }
    return new SubscribeAnnounces(requestId, trackNamespacePrefix, parameters)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('SubscribeAnnounces', () => {
    test('roundtrip', () => {
      const requestId = 241421n
      const trackNamespacePrefix = Tuple.fromUtf8Path('pre/fix/me')
      const parameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Roggan?!')),
      ]
      const msg = new SubscribeAnnounces(requestId, trackNamespacePrefix, parameters)
      const frozen = msg.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.SubscribeAnnounces))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = SubscribeAnnounces.parsePayload(frozen)
      expect(deserialized.requestId).toBe(msg.requestId)
      expect(deserialized.trackNamespacePrefix.equals(msg.trackNamespacePrefix)).toBe(true)
      expect(deserialized.parameters).toEqual(msg.parameters)
      expect(frozen.remaining).toBe(0)
    })
    test('excess roundtrip', () => {
      const requestId = 241421n
      const trackNamespacePrefix = Tuple.fromUtf8Path('pre/fix/me')
      const parameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Roggan?!')),
      ]
      const msg = new SubscribeAnnounces(requestId, trackNamespacePrefix, parameters)
      const serialized = msg.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.SubscribeAnnounces))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = SubscribeAnnounces.parsePayload(frozen)
      expect(deserialized.requestId).toBe(msg.requestId)
      expect(deserialized.trackNamespacePrefix.equals(msg.trackNamespacePrefix)).toBe(true)
      expect(deserialized.parameters).toEqual(msg.parameters)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })
    test('partial message', () => {
      const requestId = 241421n
      const trackNamespacePrefix = Tuple.fromUtf8Path('pre/fix/me')
      const parameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Roggan?!')),
      ]
      const msg = new SubscribeAnnounces(requestId, trackNamespacePrefix, parameters)
      const serialized = msg.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        SubscribeAnnounces.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/subscribe_announces.rs
use super::constant::ControlMessageType;
use super::control_message::ControlMessageTrait;
use crate::model::common::pair::KeyValuePair;
use crate::model::common::tuple::Tuple;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct SubscribeAnnounces {
  pub request_id: u64,
  pub track_namespace_prefix: Tuple,
  pub parameters: Vec<KeyValuePair>,
}

impl SubscribeAnnounces {
  pub fn new(
    request_id: u64,
    track_namespace_prefix: Tuple,
    parameters: Vec<KeyValuePair>,
  ) -> Self {
    Self {
      request_id,
      track_namespace_prefix,
      parameters,
    }
  }
}

impl ControlMessageTrait for SubscribeAnnounces {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::SubscribeAnnounces)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;
    payload.extend_from_slice(&self.track_namespace_prefix.serialize()?);

    payload.put_vi(self.parameters.len())?;
    for param in &self.parameters {
      payload.extend_from_slice(&param.serialize()?);
    }

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "Announce::serialize",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;
    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);
    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;
    let track_namespace_prefix = Tuple::deserialize(payload)?;

    let param_count_u64 = payload.get_vi()?;
    let param_count: usize =
      param_count_u64
        .try_into()
        .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
          context: "SubscribeAnnounces::deserialize(param_count)",
          from_type: "u64",
          to_type: "usize",
          details: e.to_string(),
        })?;

    let mut parameters = Vec::with_capacity(param_count);
    for _ in 0..param_count {
      let param = KeyValuePair::deserialize(payload)?;
      parameters.push(param);
    }

    Ok(Box::new(SubscribeAnnounces {
      request_id,
      track_namespace_prefix,
      parameters,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::SubscribeAnnounces
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let request_id = 241421;
    let track_namespace_prefix = Tuple::from_utf8_path("pre/fix/me");
    let parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Roggan?!")).unwrap(),
    ];
    let subscribe_announces = SubscribeAnnounces {
      request_id,
      track_namespace_prefix,
      parameters,
    };

    let mut buf = subscribe_announces.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::SubscribeAnnounces as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = SubscribeAnnounces::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe_announces);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let request_id = 241421;
    let track_namespace_prefix = Tuple::from_utf8_path("pre/fix/me");
    let parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Roggan?!")).unwrap(),
    ];
    let subscribe_announces = SubscribeAnnounces {
      request_id,
      track_namespace_prefix,
      parameters,
    };

    let serialized = subscribe_announces.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::SubscribeAnnounces as u64);
    let msg_length = buf.get_u16();

    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = SubscribeAnnounces::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe_announces);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let request_id = 241421;
    let track_namespace_prefix = Tuple::from_utf8_path("pre/fix/me");
    let parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Roggan?!")).unwrap(),
    ];
    let subscribe_announces = SubscribeAnnounces {
      request_id,
      track_namespace_prefix,
      parameters,
    };

    let mut buf = subscribe_announces.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::SubscribeAnnounces as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = SubscribeAnnounces::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
